### PR TITLE
relax .tar.gz files Content-Type check

### PR DESCRIPTION
**Pull request by Ruminateer**
**What this PR does / why we need it**:
When `storageUri` is a URI with a suffix of `.tar.gz` or `.tgz`, the current Kfserving requires `response.header.Content-Type=application/x-tar`. However, [IANA didn't define a MIME type for .tar.gz files](https://www.iana.org/assignments/media-types/media-types.xhtml), and many sites are using different `Content-Type`. Therefore, I opened the pull request to allow some other `Content-Type` for `.tar.gz` files.

Ref
- https://www.iana.org/assignments/media-types/media-types.xhtml
- https://en.wikipedia.org/wiki/List_of_archive_formats
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types
- https://github.com/python/cpython/blob/3.9/Lib/mimetypes.py
- https://stackoverflow.com/questions/28396703/mime-type-for-tgz
- https://superuser.com/questions/901962/what-is-the-correct-mime-type-for-a-tar-gz-file

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note
Do I need to write release notes?
```

### CODE REVIEW

**Comment by Ruminateer** [comment]
@googlebot I signed it!

**Comment by Ruminateer** [comment]
fixed flake8 errors

**Comment by yuzisun** [comment]
@Ruminateer Could you help add a test [here](https://github.com/kubeflow/kfserving/blob/master/python/kfserving/test/test_storage.py) ?

**Comment by Ruminateer** [comment]
> @Ruminateer Could you help add a test [here](https://github.com/kubeflow/kfserving/blob/master/python/kfserving/test/test_storage.py) ?

Added, hope how I did it looks good to you 🤨

File: python/kfserving/test/test_storage.py [hunk 1]
```diff
@@ -94,61 +96,62 @@ def test_https_uri_path(_):
...
+    def test_fail(_):
+        with pytest.raises(expected_error):
+            kfserving.Storage.download(uri)
+
+    if expected_error is None:
+        return test_success
+    return test_fail
+
+
+test_http_uri_path_with_xtar = gen_download_tests(HTTPS_URI_TARGZ,
```
Comment (yuzisun) [inline python/kfserving/test/test_storage.py:116]: I think we can use `pytest.mark.parametrize` here for different combinations of HTTP_URI and HttpResponse ?

**Comment by Ruminateer** [comment]
Used `pytest.mark.parametrize`.

Do you want me to use [it](https://github.com/Ruminateer/kfserving/blob/06f6dfcf38d75dbbf963f93685692e0ac8660223/python/kfserving/test/test_storage.py#L116-L130) for [these 2 tests](https://github.com/Ruminateer/kfserving/blob/06f6dfcf38d75dbbf963f93685692e0ac8660223/python/kfserving/test/test_storage.py#L79-L96)? I haven't done so, since technically they are not equivalent.

File: pkg/agent/storage/https.go [hunk 2]
```diff
@@ -99,7 +99,8 @@ func (h *HTTPSDownloader) Download(client http.Client) error {
 		if err := extractZipFiles(resp.Body, fileDirectory); err != nil {
 			return err
 		}
-	} else if strings.Contains(contentType, "application/x-tar") || strings.Contains(contentType, "application/gzip") {
+	} else if strings.Contains(contentType, "application/x-tar") || strings.Contains(contentType, "application/x-gtar") ||
+		strings.Contains(contentType, "application/x-gzip") || strings.Contains(contentType, "application/gzip") {
```
Comment (yuzisun) [inline pkg/agent/storage/https.go:103]: I think this condition can be simplified to 
```python
tar_content_types = ["application/x-tar", "application/x-gtar", "application/x-gzip"]
else if any(ct in contentType for ct in tar_content_types):
```

File: python/kfserving/kfserving/storage.py [hunk 3]
```diff
@@ -253,8 +253,10 @@ def _download_from_uri(uri, out_dir=None):
                     .startswith('application/zip'):
                 raise RuntimeError("URI: %s did not respond with \'Content-Type\': \'application/zip\'" % uri)
             if mimetype == 'application/x-tar' and not response.headers.get('Content-Type', '')\
-                    .startswith('application/x-tar'):
-                raise RuntimeError("URI: %s did not respond with \'Content-Type\': \'application/x-tar\'" % uri)
+                    .startswith(('application/x-tar', 'application/x-gtar', 'application/x-gzip', 'application/gzip')):
```
Comment (yuzisun) [inline python/kfserving/kfserving/storage.py:256]: prefer using a variable for the content type list so we do not have to repeat in a lot of the places

File: python/kfserving/kfserving/storage.py [hunk 4]
```diff
@@ -253,8 +253,10 @@ def _download_from_uri(uri, out_dir=None):
...
+                raise RuntimeError("URI: %s did not respond with \'Content-Type\': "
+                                   "\'application/x-tar\', \'application/x-gtar\', "
```
Comment (yuzisun) [inline python/kfserving/kfserving/storage.py:258]: same here we can use the content type list variable

File: python/kfserving/test/test_storage.py [hunk 5]
```diff
@@ -94,61 +96,38 @@ def test_https_uri_path(_):
...
+    (HTTPS_URI_TARGZ, MockHttpResponse(200, FILE_TAR_GZ_RAW, 'application/x-gzip'), None),
+    (HTTPS_URI_TARGZ, MockHttpResponse(200, FILE_TAR_GZ_RAW, 'application/gzip'), None),
+    (HTTPS_URI_TARGZ_WITH_QUERY, MockHttpResponse(200, FILE_TAR_GZ_RAW, 'application/x-tar'), None),
+    (HTTPS_URI_TARGZ_WITH_QUERY, MockHttpResponse(200, FILE_TAR_GZ_RAW, 'application/x-gtar'), None),
+    (HTTPS_URI_TARGZ_WITH_QUERY, MockHttpResponse(200, FILE_TAR_GZ_RAW, 'application/x-gzip'), None),
+    (HTTPS_URI_TARGZ_WITH_QUERY, MockHttpResponse(200, FILE_TAR_GZ_RAW, 'application/gzip'), None),
+    ('https://foo.bar/model.zip', MockHttpResponse(200, FILE_ZIP_RAW, 'application/zip'), None),
+    ('https://foo.bar/model.zip?foo=bar', MockHttpResponse(200, FILE_ZIP_RAW, 'application/zip'), None),
+    ('https://theabyss.net/model.joblib', MockHttpResponse(404), RuntimeError),
+    ('https://some.site.com/test.model', MockHttpResponse(status_code=200, content_type='text/html'), RuntimeError),
```
Comment (yuzisun) [inline python/kfserving/test/test_storage.py:111]: If the response is 200, the exception should be None right?

File: python/kfserving/test/test_storage.py [hunk 6]
```diff
@@ -94,61 +96,38 @@ def test_https_uri_path(_):
...
+    ('https://foo.bar/test/', MockHttpResponse(200), ValueError),
```
Comment (yuzisun) [inline python/kfserving/test/test_storage.py:112]: same here

File: python/kfserving/test/test_storage.py (hunk 5 above)
Comment (Ruminateer) [reply python/kfserving/test/test_storage.py:111]: These are not new tests. These are equivalent to the [original tests](https://github.com/kubeflow/kfserving/blob/a61d146b5c46706b9e9be85bf6b169262a122c95/python/kfserving/test/test_storage.py#L140-L151). The original tests expect the code to throw exceptions on bad `content_type`/bad URL even with a 200.

File: python/kfserving/test/test_storage.py (hunk 6 above)
Comment (Ruminateer) [reply python/kfserving/test/test_storage.py:112]: same as above

File: pkg/agent/storage/https.go (hunk 2 above)
Comment (Ruminateer) [reply pkg/agent/storage/https.go:103]: I don't think Go supports such syntax...

Comment (yuzisun) [reply pkg/agent/storage/https.go:103]: oops, sorry was thinking about storage initializer..

**Comment by yuzisun** [comment]
Thanks @Ruminateer !

/lgtm
/approve

**Comment by haijohn** [comment]
can you also extend  **application/zip**  to **[application/zip, application/x-zip-compressed, application/zip-compressed]**
Becase I encountered an error when download a zip file using http, and I found out the actual Content-Type is **application/x-zip-compressed** instead of **application/zip**. Please see this doc https://mimetype.io/application/zip

**Comment by vdanyliv** [comment]
@Ruminateer is there any possibility to unzip from S3?

**Pull request merged**